abb_hardware_interface/include/abb_hardware_interface/rws_service_provider_ros.hpp

```diff
@@ -0,0 +1,230 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ***********************************************************************************************************************
+ */
+//
+// This file is a modified copy from
+// https://github.com/ros-industrial/abb_robot_driver/tree/master/abb_rws_service_provider/include/abb_rws_service_provider/
+// https://github.com/ros-industrial/abb_robot_driver/tree/master/abb_rws_state_publisher/include/abb_rws_state_publisher/
+
+#ifndef ABB_HARDWARE_INTERFACE__RWS_SERCIVE_PROVIDER_ROS_HPP_
+#define ABB_HARDWARE_INTERFACE__RWS_SERCIVE_PROVIDER_ROS_HPP_
+
+#include "abb_hardware_interface/rws_client.hpp"
+
+// SYSMTEM
+#include <string>
+#include <vector>
+
+// ROS
+#include <rclcpp/rclcpp.hpp>
+
+// ABB MSG
+#include <abb_rapid_sm_addin_msgs/msg/runtime_state.hpp>
+#include <abb_robot_msgs/msg/system_state.hpp>
+
+// ABB SRV
+#include <abb_rapid_sm_addin_msgs/srv/get_egm_settings.hpp>
+#include <abb_rapid_sm_addin_msgs/srv/set_egm_settings.hpp>
+#include <abb_rapid_sm_addin_msgs/srv/set_rapid_routine.hpp>
+#include <abb_rapid_sm_addin_msgs/srv/set_sg_command.hpp>
+#include <abb_robot_msgs/srv/get_file_contents.hpp>
+#include <abb_robot_msgs/srv/get_io_signal.hpp>
+#include <abb_robot_msgs/srv/get_rapid_bool.hpp>
+#include <abb_robot_msgs/srv/get_rapid_dnum.hpp>
+#include <abb_robot_msgs/srv/get_rapid_num.hpp>
+#include <abb_robot_msgs/srv/get_rapid_string.hpp>
+#include <abb_robot_msgs/srv/get_rapid_symbol.hpp>
+#include <abb_robot_msgs/srv/get_robot_controller_description.hpp>
+#include <abb_robot_msgs/srv/get_speed_ratio.hpp>
+#include <abb_robot_msgs/srv/set_file_contents.hpp>
+#include <abb_robot_msgs/srv/set_io_signal.hpp>
+#include <abb_robot_msgs/srv/set_rapid_bool.hpp>
+#include <abb_robot_msgs/srv/set_rapid_dnum.hpp>
+#include <abb_robot_msgs/srv/set_rapid_num.hpp>
+#include <abb_robot_msgs/srv/set_rapid_string.hpp>
+#include <abb_robot_msgs/srv/set_rapid_symbol.hpp>
+#include <abb_robot_msgs/srv/set_speed_ratio.hpp>
+#include <abb_robot_msgs/srv/trigger_with_result_code.hpp>
+
+namespace abb_rws_client {
+
+class RWSServiceProviderROS : RWSClient {
+ public:
+  RWSServiceProviderROS(const rclcpp::Node::SharedPtr& node, const std::string& robot_ip, unsigned short robot_port);
+
+ private:
+  void system_state_callback(const abb_robot_msgs::msg::SystemState& msg);
+
+  void runtime_state_callback(const abb_rapid_sm_addin_msgs::msg::RuntimeState& msg);
+
+  bool get_file_contents(const abb_robot_msgs::srv::GetFileContents::Request::SharedPtr req,
```

Comment:
There are a lot of methods here, it may be prudent to subdivide them with comments grouping them similarly to the abb_ros_driver files (ex. add a comment saying core RWS services methods start here, verification methods at L#191, etc.)

Reply:
Ok, good idea. I was just wondering how to do it in the most readable way.